lib/wallet.js

```diff
@@ -0,0 +1,106 @@
+import { array, object, string } from 'yup'
+import { autowithdrawSchemaMembers, hexOrBase64Validator, lightningAddressValidator } from '@/lib/validate'
+import { TOR_REGEXP } from '@/lib/url'
+import { B64_URL_REGEX } from '@/lib/format'
+
+export function generateResolverName (walletField) {
+  const capitalized = walletField[0].toUpperCase() + walletField.slice(1)
+  return `upsertWallet${capitalized}`
+}
+
+export function generateSchema (wallet) {
```

Comment:
It looks like doing this will require us creating our own language *on top* of Yup. This won't block deployment, but I don't think we want to do this because we'll have to document it and Yup/Formik is always going to have better docs.

Formik supports [two methods of validation](https://formik.org/docs/guides/validation). So I think the move is going to be letting wallet implementers write one or the other:
1. some function that we validate with (where they can do regex or whatever) and throw errors in Formik's expected shape
2. provide their own Yup schema and we can just link yup's docs in our own docs

Reply:
Ahh, that's what you meant. I should have asked for clarification since I also wondered if this is still too depended on Yup. Didn't think about native Formik form-level validation (or field-level). Sorry for the miscommunication!

I'll revert my changes about `generateSchema` and do the following: wallet definitions can contain either `formikValidate` or `yupSchema` for validation before submit, `validateOnSubmit` which is the function that we call ourselves during submit on the client via `wallet?.validate` and `validateOnServer` which runs on the server in `upsertWallet` and thus can actually create test invoices. This is therefore also related to this TODO:

> - [ ] replace `testConnect` with `validate` so we don't have two functions for validation

So I will not replace `testConnect` with `validate` but rename them: `testConnect` to `validateOnServer` and `validate` to `validateOnSubmit`. You mentioned to only have one `validate` since both do basically the same but I think it's better to keep them separate since when we have wallets that do both, we might want to be able to run both. Also, the new names make it clear where they run whereas with only one function, it would depend on the wallet context: is it a wallet stored on the client or server?[^1] But as mentioned, we will lose this context when wallets are stored on both.[^2]

[^1]: assuming I didn't get you wrong again
[^2]: until we store encrypted spending permissions on the server but even then, we might want to run some validation during submit before the config hits the server which we don't want to run for each field change via `formikValidate` or `yupSchema`.

Reply:
`validateOnClient` might make it more consistent

Reply:
Also, it might be worth using `testConnectClient` and `testConnectServer` instead because we have both connection validation AND field validation.

> formikValidate or yupSchema for validation before submit

It might be worth giving it one name `fieldValidation` and then checking for the type returned, e.g. `typeof fieldValidation === 'function'` or `typeof fieldValidation === 'YupSchema'` or something like that.

Reply:
Mhh, as implemented in 86a630f2, we support both at the same time so it's actually not an XOR even though in practice, it's probably not useful to validate a wallet using both. I thought `Formik` might not accept both params (`validate` and `validationSchema`) at the same time but it does.

I tested the code using this patch:

```diff
diff --git a/wallets/lightning-address/index.js b/wallets/lightning-address/index.js
index 51dd665f..9ce4fc91 100644
--- a/wallets/lightning-address/index.js
+++ b/wallets/lightning-address/index.js
@@ -18,6 +18,10 @@ export const card = {
   badges: ['receive only', 'non-custodialish']
 }

+export const formikValidate = ({ address }) => {
+  return { address: 'test error' }
+}
+
 export const yupSchema = lnAddrAutowithdrawSchema

 export const walletType = 'LIGHTNING_ADDRESS'
```

and to test the validation call on the server in `upsertWallet`, I commented out `Formik.validate`:

```diff
diff --git a/pages/settings/wallets/[wallet].js b/pages/settings/wallets/[wallet].js
index c9fd8f91..442a74ff 100644
--- a/pages/settings/wallets/[wallet].js
+++ b/pages/settings/wallets/[wallet].js
@@ -40,7 +40,7 @@ export default function WalletSettings () {
       {!wallet.walletType && <WalletSecurityBanner />}
       <Form
         initial={initial}
-        validate={wallet.formikValidate}
+        // validate={wallet.formikValidate}
         schema={wallet.yupSchema}
         onSubmit={async ({ amount, ...values }) => {
           try {
```

But I'll look into using `typeof` so we can use same exported property for both for better naming.[^1]

[^1]: [`Formik` does support functions for `validationSchema`](https://formik.org/docs/api/formik#validationschema-schema----schema), too but I think we don't need to support that for wallet definitions. I don't think we need to support functions that return yup schemas there.